src/elbo.jl

```diff
@@ -0,0 +1,247 @@
+"Likelihoods which take a scalar as input and return a scalar."
+ScalarLikelihood = Union{
+    BernoulliLikelihood,
+    PoissonLikelihood,
+    GaussianLikelihood,
+    ExponentialLikelihood,
+    GammaLikelihood
+}
+
+
+abstract type ExpectationMethod end
+struct Default <: ExpectationMethod end
+struct Analytic <: ExpectationMethod end
+
+struct Quadrature <: ExpectationMethod
+    n_points::Int
+end
+Quadrature() = Quadrature(20)
+
+struct MonteCarlo <: ExpectationMethod
+    n_samples::Int
+end
+MonteCarlo() = MonteCarlo(20)
+
+"""
+    elbo(fx::FiniteGP, y::AbstractVector{<:Real}, fz::FiniteGP, q::AbstractMvNormal; n_data=length(y), method=Default())
+
+Compute the Evidence Lower BOund from [1] for the process `fx.f` where `y` are
+observations of `fx`, pseudo-inputs are given by `z = fz.x` and `q(u)` is a
+variational distribution over inducing points `u = f(z)`.
+
+`method` selects which method is used to calculate the expected loglikelihood in
+the ELBO. The options are: `Default()`, `Analytic()`, `Quadrature()` and
+`MonteCarlo()`. For likelihoods with an analytic solution, `Default()` uses this
+exact solution. If there is no such solution, `Default()` either uses
+`Quadrature()` or `MonteCarlo()`, depending on the likelihood.
+
+N.B. the observation noise `fx.Σy` is assumed to be homoscedastic and
+uncorrelated - i.e. only `fx.Σy[1]` is used.
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+function AbstractGPs.elbo(
+    fx::FiniteGP{<:AbstractGP, <:AbstractVector, <:Diagonal{<:Real, <:Fill}},
+    y::AbstractVector{<:Real},
+    fz::FiniteGP,
+    q::AbstractMvNormal;
```

Comment:
does this count as type piracy?

maybe we should leave this for separate discussion in an issue / future refactoring, but I'm wondering about whether it'd be helpful to have an object that contains everything needed for p(f | u) [for u=f(Z) that might be `fz` already, but might need to be something different for interdomain inducing variables] as well as q(u) = `q`]

Reply:
@willtebbutt @theogf

Reply:
I think refactoring this a little to put more information in a single object would be a good idea, but I'm happy for this to go in as-is for now.